src/rest.rs

```diff
@@ -36,29 +37,71 @@ impl Client {
         Ok(Client {
             config,
             address,
-            token: token,
+            token,
             rest: client,
         })
     }
 
-    pub async fn req<T: Serialize + ?Sized>(
+    pub async fn req<T: Serialize + ?Sized, Q: Serialize + ?Sized>(
         &self,
         url: String,
         method: Method,
         body: Option<&T>,
+        query: Option<&Q>,
+        headers: Option<HeaderMap>,
     ) -> Result<Response, BrigadeError> {
-        let token = self.token.clone().unwrap();
-        let req = self.rest.request(method, &url).bearer_auth(token);
-        let res = match body {
-            Some(body) => req.json(body).send().await?,
-            None => req.send().await?,
-        };
+        let req = self.create_req(url, method, body, query, headers);
 
+        let res = req.send().await?;
         Ok(res)
     }
+
+    pub async fn req_with_basic_auth<T: Serialize + ?Sized, Q: Serialize + ?Sized>(
+        &self,
+        url: String,
+        method: Method,
+        body: Option<&T>,
+        query: Option<&Q>,
+        headers: Option<HeaderMap>,
+        user: String,
+        pwd: Option<String>,
+    ) -> Result<Response, BrigadeError> {
+        let req = self.create_req(url, method, body, query, headers);
+        let res = req.basic_auth(user, pwd).send().await?;
+        Ok(res)
+    }
+
+    fn create_req<T: Serialize + ?Sized, Q: Serialize + ?Sized>(
+        &self,
+        url: String,
+        method: Method,
+        body: Option<&T>,
+        query: Option<&Q>,
```

Comment:
`Q: Serialize + ?Sized` is the type constrained that `reqwest` introduces for query parameters.